File: python/tests/test_stack.py
"""Tests for Stack implementation."""

import pytest
import numpy as np
import threading
from zeroipc import Memory, Stack


class TestStack:
    """Test Stack functionality."""
    
    def setup_method(self):
        """Clean up before each test."""
        try:
            Memory.unlink("/test_stack")
        except:
            pass
    
    def teardown_method(self):
        """Clean up after each test."""
        try:
            Memory.unlink("/test_stack")
        except:
            pass
    
    def test_create_and_basic_ops(self):
        """Test stack creation and basic operations."""
        mem = Memory("/test_stack", size=1024*1024)
        stack = Stack(mem, "int_stack", capacity=100, dtype=np.int32)
        
        assert stack.empty()
        assert not stack.full()
        assert stack.size() == 0
        assert len(stack) == 0
        
        # Push some values
        assert stack.push(10)
        assert stack.push(20)
        assert stack.push(30)
        
        assert not stack.empty()
        assert stack.size() == 3
        assert bool(stack)  # Should be truthy when not empty
        
        # Check top
        assert stack.top() == 30
        assert stack.size() == 3  # Top doesn't remove
        
        # Pop values (LIFO order)
        assert stack.pop() == 30
        assert stack.pop() == 20
        assert stack.pop() == 10
        
        assert stack.empty()
        assert stack.pop() is None
        assert stack.top() is None
    
    def test_full_stack(self):
        """Test behavior when stack is full."""
        mem = Memory("/test_stack", size=1024*1024)
        stack = Stack(mem, "small_stack", capacity=3, dtype=np.int32)
        
        assert stack.push(1)
        assert stack.push(2)
        assert stack.push(3)
        assert not stack.push(4)  # Should fail - stack full
        
        assert stack.full()
        
        stack.pop()
        assert not stack.full()
        assert stack.push(4)
    
    def test_open_existing(self):
        """Test opening an existing stack."""
        mem = Memory("/test_stack", size=1024*1024)
        
        # Create and populate stack
        stack1 = Stack(mem, "double_stack", capacity=50, dtype=np.float64)
        stack1.push(3.14)
        stack1.push(2.71)
        stack1.push(1.41)
        
        # Open same stack
        stack2 = Stack(mem, "double_stack", dtype=np.float64)
        assert stack2.size() == 3
        
        # Pop in LIFO order
        assert pytest.approx(stack2.pop()) == 1.41
        assert pytest.approx(stack2.pop()) == 2.71
        assert pytest.approx(stack2.pop()) == 3.14
    
    def test_struct_type(self):
        """Test stack with structured dtype."""
        mem = Memory("/test_stack", size=1024*1024)
        
        # Define a structured dtype
        point_dtype = np.dtype([('x', 'f4'), ('y', 'f4'), ('z', 'f4')])
        stack = Stack(mem, "point_stack", capacity=10, dtype=point_dtype)
        
        # Push structured data
        point1 = np.array((1.0, 2.0, 3.0), dtype=point_dtype)
        point2 = np.array((4.0, 5.0, 6.0), dtype=point_dtype)
        
        assert stack.push(point1)
        assert stack.push(point2)
        
        # Pop and verify (LIFO)
        p = stack.pop()
        assert p['x'] == 4.0
        assert p['y'] == 5.0
        assert p['z'] == 6.0
    
    def test_concurrent_push_pop(self):
        """Test concurrent push/pop operations."""
        mem = Memory("/test_stack", size=10*1024*1024)
        stack = Stack(mem, "concurrent_stack", capacity=10000, dtype=np.int32)
        
        num_threads = 4
        items_per_thread = 1000
        
        def worker(thread_id):
            # Push phase
            for i in range(items_per_thread):
                value = thread_id * items_per_thread + i
                while not stack.push(value):
                    pass  # Spin until successful
            
            # Pop phase (half the items)
            for i in range(items_per_thread // 2):
                while stack.pop() is None:
                    if stack.empty():
                        break
        
        # Run workers in parallel
        threads = []
        for t in range(num_threads):
            thread = threading.Thread(target=worker, args=(t,))
            threads.append(thread)
            thread.start()
        
        for thread in threads:
            thread.join()
        
        # Should have half the items left
        assert stack.size() == num_threads * items_per_thread // 2
        
        # Pop remaining
        count = 0
        while stack.pop() is not None:
            count += 1
        assert count == num_threads * items_per_thread // 2
        assert stack.empty()

File: cpp/tests/test_stack.cpp
#include <gtest/gtest.h>
#include <zeroipc/memory.h>
#include <zeroipc/stack.h>
#include <thread>
#include <vector>

using namespace zeroipc;

class StackTest : public ::testing::Test {
protected:
    void SetUp() override {
        Memory::unlink("/test_stack");
    }
    
    void TearDown() override {
        Memory::unlink("/test_stack");
    }
};

TEST_F(StackTest, CreateAndBasicOps) {
    Memory mem("/test_stack", 1024*1024);
    Stack<int> stack(mem, "int_stack", 100);
    
    EXPECT_TRUE(stack.empty());
    EXPECT_FALSE(stack.full());
    EXPECT_EQ(stack.size(), 0);
    EXPECT_EQ(stack.capacity(), 100);
    
    // Push some values
    EXPECT_TRUE(stack.push(10));
    EXPECT_TRUE(stack.push(20));
    EXPECT_TRUE(stack.push(30));
    
    EXPECT_FALSE(stack.empty());
    EXPECT_EQ(stack.size(), 3);
    
    // Check top
    auto top = stack.top();
    ASSERT_TRUE(top.has_value());
    EXPECT_EQ(*top, 30);
    
    // Pop values (LIFO order)
    auto val = stack.pop();
    ASSERT_TRUE(val.has_value());
    EXPECT_EQ(*val, 30);
    
    val = stack.pop();
    ASSERT_TRUE(val.has_value());
    EXPECT_EQ(*val, 20);
    
    val = stack.pop();
    ASSERT_TRUE(val.has_value());
    EXPECT_EQ(*val, 10);
    
    EXPECT_TRUE(stack.empty());
    EXPECT_FALSE(stack.pop().has_value());
    EXPECT_FALSE(stack.top().has_value());
}

TEST_F(StackTest, FullStack) {
    Memory mem("/test_stack", 1024*1024);
    Stack<int> stack(mem, "small_stack", 3);
    
    EXPECT_TRUE(stack.push(1));
    EXPECT_TRUE(stack.push(2));
    EXPECT_TRUE(stack.push(3));
    EXPECT_FALSE(stack.push(4)); // Should fail - stack full
    
    EXPECT_TRUE(stack.full());
    
    stack.pop();
    EXPECT_FALSE(stack.full());
    EXPECT_TRUE(stack.push(4));
}

TEST_F(StackTest, OpenExisting) {
    Memory mem("/test_stack", 1024*1024);
    
    {
        Stack<double> stack1(mem, "double_stack", 50);
        stack1.push(3.14);
        stack1.push(2.71);
        stack1.push(1.41);
    }
    
    Stack<double> stack2(mem, "double_stack");
    EXPECT_EQ(stack2.capacity(), 50);
    EXPECT_EQ(stack2.size(), 3);
    
    EXPECT_DOUBLE_EQ(*stack2.pop(), 1.41);
    EXPECT_DOUBLE_EQ(*stack2.pop(), 2.71);
    EXPECT_DOUBLE_EQ(*stack2.pop(), 3.14);
}

TEST_F(StackTest, StructType) {
    struct Point {
        float x, y, z;
    };
    
    Memory mem("/test_stack", 1024*1024);
    Stack<Point> stack(mem, "point_stack", 10);
    
    Point p1{1.0f, 2.0f, 3.0f};
    Point p2{4.0f, 5.0f, 6.0f};
    
    EXPECT_TRUE(stack.push(p1));
    EXPECT_TRUE(stack.push(p2));
    
    auto p = stack.pop();
    ASSERT_TRUE(p.has_value());
    EXPECT_FLOAT_EQ(p->x, 4.0f);
    EXPECT_FLOAT_EQ(p->y, 5.0f);
    EXPECT_FLOAT_EQ(p->z, 6.0f);
}

TEST_F(StackTest, ConcurrentPushPop) {
    Memory mem("/test_stack", 10*1024*1024);
    Stack<int> stack(mem, "concurrent_stack", 10000);
    
    const int num_threads = 4;
    const int items_per_thread = 1000;
    std::vector<std::thread> threads;
    
    // Multiple threads pushing and popping
    for (int t = 0; t < num_threads; t++) {
        threads.emplace_back([&, t]() {
            // Push phase
            for (int i = 0; i < items_per_thread; i++) {
                int value = t * items_per_thread + i;
                while (!stack.push(value)) {
                    std::this_thread::yield();
                }
            }
            
            // Pop phase
            for (int i = 0; i < items_per_thread / 2; i++) {
                while (!stack.pop()) {
                    std::this_thread::yield();
                }
            }
        });
    }
    
    for (auto& t : threads) t.join();
    
    // Should have half the items left
    EXPECT_EQ(stack.size(), num_threads * items_per_thread / 2);
    
    // Pop remaining
    int count = 0;
    while (stack.pop()) {
        count++;
    }
    EXPECT_EQ(count, num_threads * items_per_thread / 2);
    EXPECT_TRUE(stack.empty());
}